airbyte-webapp/src/views/Connection/ConnectionForm/calculateInitialCatalog.ts

```diff
@@ -0,0 +1,108 @@
+import {
+  AirbyteStreamConfiguration,
+  DestinationSyncMode,
+  SyncMode,
+  SyncSchema,
+  SyncSchemaStream,
+} from "core/domain/catalog";
+
+const getDefaultCursorField = (streamNode: SyncSchemaStream): string[] => {
+  if (streamNode.stream.defaultCursorField.length) {
+    return streamNode.stream.defaultCursorField;
+  }
+  return streamNode.config.cursorField;
+};
+
+const verifySupportedSyncModes = (streamNode: SyncSchemaStream): SyncSchemaStream => {
+  const {
+    stream: { supportedSyncModes },
+  } = streamNode;
+
+  if (supportedSyncModes?.length) {
+    return streamNode;
+  }
+  return { ...streamNode, stream: { ...streamNode.stream, supportedSyncModes: [SyncMode.FullRefresh] } };
+};
+
+const verifyConfigCursorField = (streamNode: SyncSchemaStream): SyncSchemaStream => {
+  const { config } = streamNode;
+
+  return {
+    ...streamNode,
+    config: {
+      ...config,
+      cursorField: config.cursorField?.length ? config.cursorField : getDefaultCursorField(streamNode),
+    },
+  };
+};
+
+const getOptimalSyncMode = (
+  streamNode: SyncSchemaStream,
+  supportedDestinationSyncModes: DestinationSyncMode[]
+): SyncSchemaStream => {
+  const updateStreamConfig = (
+    config: Pick<AirbyteStreamConfiguration, "syncMode" | "destinationSyncMode">
+  ): SyncSchemaStream => ({
+    ...streamNode,
+    config: { ...streamNode.config, ...config },
+  });
+
+  const {
+    stream: { supportedSyncModes, sourceDefinedCursor },
+  } = streamNode;
+
+  if (
+    supportedSyncModes.includes(SyncMode.Incremental) &&
+    supportedDestinationSyncModes.includes(DestinationSyncMode.Dedupted) &&
+    sourceDefinedCursor
+  ) {
+    return updateStreamConfig({
+      syncMode: SyncMode.Incremental,
+      destinationSyncMode: DestinationSyncMode.Dedupted,
+    });
+  }
+
+  if (supportedDestinationSyncModes.includes(DestinationSyncMode.Overwrite)) {
+    return updateStreamConfig({
+      syncMode: SyncMode.FullRefresh,
+      destinationSyncMode: DestinationSyncMode.Overwrite,
+    });
+  }
+
+  if (
+    supportedSyncModes.includes(SyncMode.Incremental) &&
+    supportedDestinationSyncModes.includes(DestinationSyncMode.Append)
+  ) {
+    return updateStreamConfig({
+      syncMode: SyncMode.Incremental,
+      destinationSyncMode: DestinationSyncMode.Append,
+    });
+  }
+
+  if (supportedDestinationSyncModes.includes(DestinationSyncMode.Append)) {
+    return updateStreamConfig({
+      syncMode: SyncMode.FullRefresh,
+      destinationSyncMode: DestinationSyncMode.Append,
+    });
+  }
+  return streamNode;
+};
+
+const calculateInitialCatalog = (
+  schema: SyncSchema,
+  supportedDestinationSyncModes: DestinationSyncMode[],
+  isEditMode?: boolean
+): SyncSchema => ({
+  streams: schema.streams.map<SyncSchemaStream>((apiNode, id) => {
+    const nodeWithId: SyncSchemaStream = { ...apiNode, id: id.toString() };
+    const nodeStream = verifySupportedSyncModes(nodeWithId);
```

Comment:
This is the only place I am still a tiny bit concerned around. This method could change the existing Sync Mode when opening a connection in edit mode, if the previously set sync mode isn't available anymore. In that case the user would see a different sync mode in the UI, than the DB actually has, and also they would not be able to save it(??) because it's still the default values of the form?

@sherifnada Is there any case were a connector (e.g. due to an update) could lose a sync mode on an existing/saved connection's stream that was earlier supported? Is this something we'd need to cater for int he UI, or should that anyway never happen?

cc @edmundito @dizel852

Reply:
If this is nothing we anticipate happening, I'd suggest we simply leave it empty and force the user to select a new one. @dizel852 Do you think we could still change that logic not to jump to full refresh overwrite if the actual configured sync doesn't exist, but leave the select box empty and invalid and force the user to select? (And ideally also add a unit test case for that scenario :D) Thanks

Reply:
@timroes Yes, we can omit 'supportedSyncMode' checking for edit mode. 
But what should we verify it in 'creation' mode? 
If we will leave 'supportedSyncMode' checking just for creation we will get a behavior:
https://www.loom.com/share/062406b7dd26417580a914ec6ffef0c6

As you can see, available sync modes rely on at least one (or default) value in  'supportedSyncMode'.

IMO One of the solutions - does not verify 'supportedSyncMode' at all,  neither in creation nor edit modes.
Do we get empty empty  'supportedSyncMode'? - Okay, that means that our connection form is invalid since there are no available sync modes. So we just do not allow the user to create a connection.

Advantages: We don't try somehow fix the backend's response. Another question is why do we get such a response?
Disadvantages: More fixes 😅 Need to update our form validations and tests.

WDYT?

Reply:
@dizel852 I agree with the part: that we should not try to fix a broken connector (like the end-2-end one) in the FE by overwriting it's available sync modes. I think we should simply not adjust them, and in this case not being able to create a connection with that connector. Otherwise we'd allow users to create a connection with a configuration the backend told us is invalid, which does not sound like a good idea.

Despite being more work I think we should make sure that we error the form in this case, i.e. allow for an `undefined` value in case there is no valid syncCode, check for that in our [validation code](https://github.com/airbytehq/airbyte/blob/master/airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx#L110), and make sure we're setting `undefined` if there's no valid configuration (either in `getOptimalSyncMode` or rather add another method call to `calculateInitialCatalog` for that.

With that said, we should still make sure our end-to-end testing connectors will not result in an invalid configuration that other connectors in practise can't end up. @sherifnada would you maybe able to take that to the connector team? Also if we change the UI here, without fixing the connectors I am not sure if we actually still can run end-to-end tests, thus it would be good if we could solve that somehow in the near future if possible :)

Reply:
> Despite being more work I think we should make sure that we error the form in this case, i.e. allow for an `undefined` value in case there is no valid syncCode, check for that in our [validation code](https://github.com/airbytehq/airbyte/blob/master/airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx#L110), and make sure we're setting `undefined` if there's no valid configuration (either in `getOptimalSyncMode` or rather add another method call to `calculateInitialCatalog` for that.

Ok, got it. Will update the implementation. 👍